db/migration/1590583070171-ExtractMapColorScale.ts

```diff
@@ -0,0 +1,138 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
+import omit from "lodash/omit"
+import clone from "lodash/clone"
+
+type MapProjection =
+    | "World"
+    | "Africa"
+    | "NorthAmerica"
+    | "SouthAmerica"
+    | "Asia"
+    | "Europe"
+    | "Oceania"
+
+interface ColorScaleConfig {
+    baseColorScheme?: string
+    colorSchemeMinValue?: number
+    colorSchemeValues: number[]
+    colorSchemeLabels: (string | undefined)[]
+    isManualBuckets?: true
+    equalSizeBins?: true
+    colorSchemeInvert?: true
+    customColorsActive?: true
+    customNumericColors: (string | undefined)[]
+    customCategoryColors: { [key: string]: string }
+    customCategoryLabels: { [key: string]: string }
+    customHiddenCategories: { [key: string]: true }
+    legendDescription?: string
+    binStepSize?: number
+}
+
+interface MapOnlyConfig {
+    variableId?: number
+    targetYear?: number
+    timeTolerance?: number
+    hideTimeline?: true
+    projection: MapProjection
+    tooltipUseCustomLabels?: true
+}
+
+type OldMapConfig = ColorScaleConfig & MapOnlyConfig
+
+type NewMapConfig = MapOnlyConfig & { colorScale: ColorScaleConfig }
+
+function extractColorScaleConfig(
+    oldConfig: OldMapConfig | undefined | null
+): NewMapConfig | undefined {
+    if (oldConfig === undefined || oldConfig === null) return undefined
+    const {
+        baseColorScheme,
+        colorSchemeMinValue,
+        colorSchemeValues,
+        colorSchemeLabels,
+        isManualBuckets,
+        equalSizeBins,
+        colorSchemeInvert,
+        customColorsActive,
+        customNumericColors,
+        customCategoryColors,
+        customCategoryLabels,
+        customHiddenCategories,
+        legendDescription,
+        binStepSize,
+        ...rest
+    } = oldConfig
+    return {
+        ...rest,
+        colorScale: {
+            baseColorScheme,
+            colorSchemeMinValue,
+            colorSchemeValues,
+            colorSchemeLabels,
+            isManualBuckets,
+            equalSizeBins,
+            colorSchemeInvert,
+            customColorsActive,
+            customNumericColors,
+            customCategoryColors,
+            customCategoryLabels,
+            customHiddenCategories,
+            legendDescription,
+            binStepSize
+        }
+    }
+}
+
+function mergeColorScaleConfig(
+    newConfig: NewMapConfig | undefined | null
+): OldMapConfig | undefined {
+    if (newConfig === undefined || newConfig === null) return undefined
+    const { colorScale } = newConfig
+    return {
+        ...omit(newConfig, "colorScale"),
+        ...colorScale
+    }
+}
+
+async function transformAllCharts(
+    queryRunner: QueryRunner,
+    getNewConfig: (oldConfig: any) => any
+) {
+    const charts = (await queryRunner.query(
+        "SELECT charts.id AS id, charts.config AS config FROM charts"
```

Comment:
Do we need to worry about "Revisions"/"Restore", or is that rarely used?

Reply:
Good point! I didn't think of that. It's rarely used as far as I'm aware, but I'll try to migrate those too

Reply:
Now added those to the migration